qiskit/transpiler/passes/scheduling/dynamical_decoupling.py

```diff
@@ -0,0 +1,219 @@
+                equal spacing will be used.
+            skip_reset_qubits (bool): if True, does not insert DD on idle
+                periods that immediately follow initialized/reset qubits (as
+                qubits in the ground state are less susceptile to decoherence).
+        """
+        super().__init__()
+        self._durations = durations
+        self._dd_sequence = dd_sequence
+        self._qubits = qubits
+        self._spacing = spacing
+        self._skip_reset_qubits = skip_reset_qubits
+
+    def run(self, dag):
+        """Run the DynamicalDecoupling pass on dag.
+
+        Args:
+            dag (DAGCircuit): a scheduled DAG.
+
+        Returns:
+            DAGCircuit: equivalent circuit with delays interrupted by DD,
+                where possible.
+
+        Raises:
+            TranspilerError: if the circuit is not mapped on physical qubits.
+        """
+        if len(dag.qregs) != 1 or dag.qregs.get("q", None) is None:
+            raise TranspilerError("DD runs on physical circuits only.")
+
+        if dag.duration is None:
+            raise TranspilerError("DD runs after circuit is scheduled.")
+
+        # we support only even number of DD pulses (or just 1, as a special case)
+        if len(self._dd_sequence) != 1:
+            if len(self._dd_sequence) % 2 != 0:
+                raise TranspilerError("DD sequence must contain an even number of gates (or 1).")
+            noop = Operator(IGate())
+            for gate in self._dd_sequence:
+                noop = noop.compose(Operator(gate))
+            if not noop.equiv(IGate()):
+                raise TranspilerError("The DD sequence does not make an identity operation.")
+
+        if self._qubits is None:
+            self._qubits = range(dag.num_qubits())
+
+        if self._spacing:
+            raise TranspilerError("only balanced spacing is supported now.")
+
+        new_dag = dag._copy_circuit_metadata()
+
+        for nd in dag.topological_op_nodes():
+            if isinstance(nd.op, Delay):
+                dag_qubit = nd.qargs[0]
+                physical_qubit = new_dag.qubits.index(dag_qubit)
+                if physical_qubit not in self._qubits:  # skip unwanted qubits
+                    new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                    continue
+
+                pred = next(dag.predecessors(nd))
+                succ = next(dag.successors(nd))
+                if self._skip_reset_qubits:  # discount initial delays
+                    if pred.type == "in" or isinstance(pred, Reset):
+                        new_dag.apply_operation_back(nd.op, nd.qargs, nd.cargs)
+                        continue
+
+                dd_sequence = []
+                dd_sequence_duration = 0
+                for g in self._dd_sequence:
+                    g.duration = self._durations.get(g, physical_qubit)
+                    dd_sequence.append(g)
+                    dd_sequence_duration += g.duration
```

Comment:
Could be memoized for each qubit

Reply:
unfortunately the `physical_qubit` changes the duration. but I did realize that I needlessly created a `dd_sequence` again which is the same as `self._dd_sequence`